src/libraries/System.Private.CoreLib/src/System/Text/NewLineUtility.cs

```diff
@@ -0,0 +1,51 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text
+{
+    internal static class NewLineUtility
+    {
+        public static int GetIndexOfFirstNewLineChar(ReadOnlySpan<char> span, out int charsToConsume)
+        {
+            // !! IMPORTANT !!
+            //
+            // We expect this method may be called with untrusted input, which means we need to bound the
+            // worst-case runtime of this method. The logic below is specially written to run in worst-case
+            // O(i) runtime, where i is the index of the first newline char in span. If there exists
+            // no newline char in span, then the runtime is O(span.Length). The reason for this is that
+            // we expect this method to be called in a loop and for the caller to slice, and this allows us
+            // bound the entire loop in worst-case O(span.Length) runtime.
+            //
+            // Importantly, this means that we cannot call MemoryExtensions.IndexOfAny, as that method
+            // is worst-case bounded by O(span.Length), not O(i). This means that when called in a loop,
+            // the worst-case runtime for the loop is O(span.Length ^ 2), which could lead to an algorithmic
+            // complexity attack.
```

Comment:
I don't understand this.  Can you elaborate?  We will always jump forward to whatever is the result of IndexOfAny, so we'd never end up looking at any position more than once.  How would this degrade to O(n^2)?

Reply:
The call to `IndexOfAny` may be _O(n)_, regardless of where the index actually occurs within the input. Imagine an input crafted such that there's a match at every index. So our first call to `IndexOfAny` will be _O(span.Length)_. Our second call might be _O(span.Length - 1)_. The third call might be _O(span.Length - 2)_. And so on.

The amortized cost between slices is _O(n)_, which puts the aggregate operation at _O(n^2)_.

Reply:
> The call to IndexOfAny may be O(n), regardless of where the index actually occurs within the input.

Why?  You're saying we might find the match at position 1 but still examine 1,000,000 more characters in the input?

Reply:
> You're saying we might find the match at position 1 but still examine 1,000,000 more characters in the input?

`IndexOfAny` (for inputs > 5 elements) has this behavior today, yes. In fact, if you replace this routine with a call to `IndexOfAny`, the unit test _EnumerateLines\_EnumerationIsNotPolynomialComplexity_ (introduced as part of this PR) will detect the _O(n^2)_ operation and fail.

We have a few options for addressing this. We can change `IndexOfAny` to no longer have _O(n)_ behavior (which may severely harm perf for other common use cases). We can yet another special-cased code path within `IndexOfAny`. These specialized code paths within `IndexOfAny` tend to be very complex and utilize significant unsafe code. I had planned on addressing the performance issue within `NewLineUtility` directly using a minimum amount of unsafe code and relying on general vectorized APIs.

Reply:
> IndexOfAny (for inputs > 5 elements) has this behavior today, yes

Sorry, I feel like I'm missing something, as what you describe sounds unshippable (yet we apparently shipped it).  You're saying that even after IndexOfAny finds a match, it continues searching the entire rest of the input for another match?

If yes, we need to fix that.

Reply:
I just looked at the implementation:
https://github.com/dotnet/runtime/blob/ebba45155160c7056faf1675448af1b75e9f4a48/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs#L597-L609
Seems the behavior you describe is in fact there and happens because we're doing a naive "first search for value 1, then if it couldn't be found search for value 2, then if it couldn't be found search for value 3, ... and so on".  Ugh.

Reply:
> IndexOfAny is used by a non-trivial amount of code to do parsing like this PR is doing. The workaround employed of not using IndexOfAny doesn't help those other uses. I think we need to fix this.

I agree that we should consider a change here. The reason I wasn't pushing the issue hard is that I couldn't find any perf-critical call sites within our SDK which provide 6+ needles. The highest we've gone in the past is 5 needles (for `File` APIs), and we have a specialized code path to handle that. As far as I know, this PR introduces the first API which needs to perform a 6-needle search.

My main concern behind making a change to `IndexOfAny` for the 6+ needle case is that it could be a significant perf regression in the case where the needles are sorted roughly in order of likelihood. We can guarantee _O(i * p)_ runtime \[where _i_ is the index of the first match in the haystack and _p_ is the number of needles being sought\], but I wonder if that "_\* p_" part is going to irritate people.

Reply:
Also, there is the whole overhead of setting up the SIMD loop on *ever single entrance* to `IndexOfAny`, but eh. Unless we often expect matches to occur near the beginning of the span this overhead might not be worth worrying about.

Reply:
Yeah, totally get there will be tradeoffs, and we may not like all of them.  But for a long search space, I think where we currently are is worse.  Thanks for thinking deeply about it and raising the issue.